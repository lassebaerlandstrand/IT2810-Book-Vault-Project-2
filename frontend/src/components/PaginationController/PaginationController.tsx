import { useSearchParams } from 'react-router-dom';
import { Center, Group, Pagination } from '@mantine/core';
import { usePaginationParams } from '@/hooks/usePagination';

type PaginationControllerProps = {
  totalBooks: number;
};

const PaginationController = ({ totalBooks }: PaginationControllerProps) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const { page, limit } = usePaginationParams();

  const handlePageChange = (newPage: number) => {
    const newParams = new URLSearchParams(searchParams.toString());
    newParams.set('page', newPage.toString());
    setSearchParams(newParams);
  };

  return (
    <>
      <Center>
        <Pagination.Root
          total={Math.ceil(totalBooks / limit)}
          value={page}
          onChange={handlePageChange}
          my="md"
        >
          <Group gap="xs">
            <Pagination.Previous aria-label="Previous Page" />
            <Pagination.Items />
            <Pagination.Next aria-label="Next Page" />
          </Group>
        </Pagination.Root>
      </Center>
    </>
  );
};

export default PaginationController;
