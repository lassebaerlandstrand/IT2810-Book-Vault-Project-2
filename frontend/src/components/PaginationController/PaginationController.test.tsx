import { fireEvent, render, screen } from '@test-utils';
import { MemoryRouter } from 'react-router-dom';
import * as reactRouterDom from 'react-router-dom'; // This was the only was to get spyOn to work with useSearchParams
import PaginationController from './PaginationController';

vi.mock('@/hooks/usePaginationParams', () => ({
  usePaginationParams: vi.fn(() => ({ page: 1, limit: 25 })),
}));

describe('PaginationController', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders the pagination component', () => {
    render(
      <MemoryRouter>
        <PaginationController totalBooks={100} />
      </MemoryRouter>
    );

    expect(screen.getByText('1')).toBeInTheDocument(); // Check that page 1 is displayed
    expect(screen.getByText('4')).toBeInTheDocument(); // Check that page 4 is displayed for totalBooks = 100 and limit = 25
  });

  test('calls setSearchParams with the correct page when the page changes', () => {
    const mockSetSearchParams = vi.fn();
    vi.spyOn(reactRouterDom, 'useSearchParams').mockReturnValue([
      new URLSearchParams(),
      mockSetSearchParams,
    ]);

    render(
      <MemoryRouter>
        <PaginationController totalBooks={100} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('2')); // Simulate clicking on page 2

    expect(mockSetSearchParams).toHaveBeenCalledWith(expect.any(URLSearchParams));
    const params = mockSetSearchParams.mock.calls[0][0];
    expect(params.get('page')).toBe('2');
  });

  test('calculates total pages correctly based on totalBooks and limit', () => {
    render(
      <MemoryRouter>
        <PaginationController totalBooks={75} />{' '}
        {/* Total 75 books with a limit of 25, should show 3 pages */}
      </MemoryRouter>
    );

    expect(screen.getByText('3')).toBeInTheDocument(); // Check that page 3 is displayed
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<PaginationController totalBooks={75} />);
    const attributesToRemove = document.body.querySelectorAll('div [id^="mantine"]'); // Because Mantine uses random ids which causes snapshots to fail
    attributesToRemove.forEach((element) => {
      element.removeAttribute('id');
      element.removeAttribute('aria-describedby');
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
