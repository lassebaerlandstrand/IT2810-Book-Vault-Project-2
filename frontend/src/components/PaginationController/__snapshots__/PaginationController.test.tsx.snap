// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PaginationController > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="m_4451eb3a mantine-Center-root"
  >
    <div
      class="m_4addd315 mantine-Pagination-root"
      style="margin-block: var(--mantine-spacing-md);"
    >
      <div
        class="m_4081bf90 mantine-Group-root"
        style="--group-gap: calc(0.5rem * var(--mantine-scale)); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
      >
        <button
          class="mantine-focus-auto m_326d024a mantine-Pagination-control m_87cf2631 mantine-UnstyledButton-root"
          data-disabled="true"
          disabled=""
          type="button"
        >
          <svg
            class="mantine-rotate-rtl"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.219 8l3.3 3.3-.943.943L5.333 8l4.243-4.243.943.943-3.3 3.3z"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          aria-current="page"
          class="mantine-focus-auto mantine-active m_326d024a mantine-Pagination-control m_87cf2631 mantine-UnstyledButton-root"
          data-active="true"
          data-with-padding="true"
          type="button"
        >
          1
        </button>
        <button
          class="mantine-focus-auto mantine-active m_326d024a mantine-Pagination-control m_87cf2631 mantine-UnstyledButton-root"
          data-with-padding="true"
          type="button"
        >
          2
        </button>
        <button
          class="mantine-focus-auto mantine-active m_326d024a mantine-Pagination-control m_87cf2631 mantine-UnstyledButton-root"
          data-with-padding="true"
          type="button"
        >
          3
        </button>
        <button
          class="mantine-focus-auto mantine-active m_326d024a mantine-Pagination-control m_87cf2631 mantine-UnstyledButton-root"
          type="button"
        >
          <svg
            class="mantine-rotate-rtl"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.781 8l-3.3-3.3.943-.943L10.667 8l-4.243 4.243-.943-.943 3.3-3.3z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
